README.md
[![Build Status](https://travis-ci.org/genio/psperl.svg?branch=master)](https://travis-ci.org/genio/psperl)

# NAME

PSPerl - Manage Strawberry Perl (Portable) installations in your $env:HOMEPATH

# SYNOPSIS

A typical install, selection, and use case would look something like this:

```PowerShell
PS C:\Users\genio> git clone https://github.com/genio/psperl.git _psperl
PS C:\Users\genio> cd _psperl
PS C:\Users\genio\_psperl> .\psperl.ps1 -Setup
PS C:\Users\genio\_psperl> .\psperl.ps1 -Init
PS C:\Users\genio\_psperl> cd ..
PS C:\Users\genio> psperl -Available
...
PS C:\Users\genio> psperl -Install perl64-5.30.0.1
PS C:\Users\genio> psperl -Switch perl64-5.30.0.1
PS C:\Users\genio> perl -v

This is perl 5, version 30, subversion 0 (v5.30.0) built for MSWin32-x86-multi-thread-64int

Copyright 1987-2019, Larry Wall

Perl may be copied only under the terms of either the Artistic License or the
GNU General Public License, which may be found in the Perl 5 source kit.

Complete documentation for Perl, including FAQ lists, should be found on
this system using "man perl" or "perldoc perl".  If you have access to the
Internet, point your browser at http://www.perl.org/, the Perl Home Page.

PS C:\Users\genio>
```

Then, every subsequent time you load up PowerShell, you can forego the installation, etc.

```PowerShell
PS C:\Users\genio> perl -v

This is perl 5, version 30, subversion 0 (v5.30.0) built for MSWin32-x86-multi-thread-64int

Copyright 1987-2019, Larry Wall

Perl may be copied only under the terms of either the Artistic License or the
GNU General Public License, which may be found in the Perl 5 source kit.

Complete documentation for Perl, including FAQ lists, should be found on
this system using "man perl" or "perldoc perl".  If you have access to the
Internet, point your browser at http://www.perl.org/, the Perl Home Page.

PS C:\Users\genio>
```

# INSTALLATION

```PowerShell
git clone https://github.com/genio/psperl.git
cd psperl
.\psperl.ps1 -Setup
```

Setting things up couldn't be easier. Basically, all you need to do is check out this repository or download it as a .zip file and extract it somewhere. Once you have it sitting in some folder, navigate to that folder in PowerShell and run the setup command: `.\psperl.ps1 -Setup`.

That will open your `$Profile` and add the following lines:
```PowerShell
# PSPerl Initialize
& C:\Users\genio\_psperl\psperl.ps1 -Init
```

That addition to your profile will handle ensuring that PSPerl is always in your PowerShell's `$env:Path` environment variable. From that point, you could just run `psperl -available` to get a list of available Portable [Strawberry Perls](http://strawberryperl.com) to install.

# DESCRIPTION

[Strawberry Perl](http://strawberryperl.com) is a Perl built to run on Windows. It's a vanilla Perl install that the Perl community has put together so that you don't have to go through the hassle of trying to get all of the prerequisites and other things necessary to build Perl yourself done.

Thank you, Strawberry Perl!

# COMMANDS

PSPerl has a few commands to help you through the process of getting a working version of Perl installed.

## Available

```PowerShell
psperl -Available # Shows you a list of all available installations
psperl -Available -major 10 # shows you a list of all Perls available on v5.10
```

The `Available` command will grab the list of releases from [Strawberry Perl's JSON file](http://strawberryperl.com/releases.json) and parse out which releases are available in Portable form -release types found [here](http://strawberryperl.com/releases.html). The information coming from Strawberry Perl's site will be cached locally on your hard drive as a JSON file that we will use from that point on so we don't hit their servers too often and to speed up your experience. This local cache will time out after an hour and a half, at which time we'll grab another copy from Strawberry's site.

It will spit out a list of all of the Perl versions you can install for use on your installation combination of Windows and PowerShell. This part is a bit different from other brew-style installers because we try to account for your `64-bit` vs `32-bit` vs `32-bit with USE_64_BIT_INT`.

### 64-bit Windows, 64-bit PowerShell

This is what you could expect to see on this platform:

```PowerShell
PS C:\Users\genio> psperl -Available
We're using a 64-bit PowerShell on a 64-bit Windows OS.
On 64-bit Windows:
    32-bit PowerShell: You can use 32-bit, 32-bit USE_64_BIT_INT Perls without problem.
                       You _can_ use 64-bit Perls, but may run into trouble. RISK
    64-bit PowerShell: You can use 32-bit, 32-bit USE_64_BIT_INT, or 64-bit Perls.
On 32-bit Windows:
    You can only use 32-bit or 32-bit with USE_64_BIT_INT Perls

The current release for each major Perl:

64-bit Perls:
   - perl64-5.30.0.1
   - perl64-5.28.2.1
   - perl64-5.26.3.1
   - perl64-5.24.4.1
   - perl64-5.22.3.1
   - perl64-5.20.3.3
   - perl64-5.18.4.1
   - perl64-5.16.3.1
   - perl64-5.14.4.1

32-bit with USE_64_BIT_INT Perls:
   - perl32w64int-5.30.0.1
   - perl32w64int-5.28.2.1
   - perl32w64int-5.26.3.1
   - perl32w64int-5.24.4.1
   - perl32w64int-5.22.3.1
   - perl32w64int-5.20.3.3
   - perl32w64int-5.18.4.1

32-bit Perls:
   - perl32-5.30.0.1
   - perl32-5.28.2.1
   - perl32-5.26.3.1
   - perl32-5.24.4.1
   - perl32-5.22.3.1
   - perl32-5.20.3.3
   - perl32-5.18.4.1
   - perl32-5.16.3.1
   - perl32-5.14.4.1
   - perl32-5.12.3.0
   - perl32-5.10.1.2

PS C:\Users\genio>
```

### 64-bit Windows, 32-bit PowerShell

This is what you could expect to see on this platform:

```PowerShell
PS C:\Windows\SysWOW64\WindowsPowerShell\v1.0> psperl -Available
We're using a 32-bit PowerShell on a 64-bit Windows OS.
On 64-bit Windows:
    32-bit PowerShell: You can use 32-bit, 32-bit USE_64_BIT_INT Perls without problem.
                       You _can_ use 64-bit Perls, but may run into trouble. RISK
    64-bit PowerShell: You can use 32-bit, 32-bit USE_64_BIT_INT, or 64-bit Perls.
On 32-bit Windows:
    You can only use 32-bit or 32-bit with USE_64_BIT_INT Perls

The current release for each major Perl:

64-bit Perls (RISKY on a 32-bit PS):
   - perl64-5.30.0.1
   - perl64-5.28.2.1
   - perl64-5.26.3.1
   - perl64-5.24.4.1
   - perl64-5.22.3.1
   - perl64-5.20.3.3
   - perl64-5.18.4.1
   - perl64-5.16.3.1
   - perl64-5.14.4.1

32-bit with USE_64_BIT_INT Perls:
   - perl32w64int-5.30.0.1
   - perl32w64int-5.28.2.1
   - perl32w64int-5.26.3.1
   - perl32w64int-5.24.4.1
   - perl32w64int-5.22.3.1
   - perl32w64int-5.20.3.3
   - perl32w64int-5.18.4.1

32-bit Perls:
   - perl32-5.30.0.1
   - perl32-5.28.2.1
   - perl32-5.26.3.1
   - perl32-5.24.4.1
   - perl32-5.22.3.1
   - perl32-5.20.3.3
   - perl32-5.18.4.1
   - perl32-5.16.3.1
   - perl32-5.14.4.1
   - perl32-5.12.3.0
   - perl32-5.10.1.2

PS C:\Windows\SysWOW64\WindowsPowerShell\v1.0>
```

### 32-bit Windows
And, finally, on this platform, you'd see:

```PowerShell
PS C:\Users\genio> psperl -Available
We're using a 32-bit PowerShell on a 32-bit Windows OS.
On 64-bit Windows:
    32-bit PowerShell: You can use 32-bit, 32-bit USE_64_BIT_INT Perls without problem.
                       You _can_ use 64-bit Perls, but may run into trouble. RISK
    64-bit PowerShell: You can use 32-bit, 32-bit USE_64_BIT_INT, or 64-bit Perls.
On 32-bit Windows:
    You can only use 32-bit or 32-bit with USE_64_BIT_INT Perls

The current release for each major Perl:

32-bit with USE_64_BIT_INT Perls:
   - perl32w64int-5.30.0.1
   - perl32w64int-5.28.2.1
   - perl32w64int-5.26.3.1
   - perl32w64int-5.24.4.1
   - perl32w64int-5.22.3.1
   - perl32w64int-5.20.3.3
   - perl32w64int-5.18.4.1

32-bit Perls:
   - perl32-5.30.0.1
   - perl32-5.28.2.1
   - perl32-5.26.3.1
   - perl32-5.24.4.1
   - perl32-5.22.3.1
   - perl32-5.20.3.3
   - perl32-5.18.4.1
   - perl32-5.16.3.1
   - perl32-5.14.4.1
   - perl32-5.12.3.0
   - perl32-5.10.1.2

PS C:\Users\genio>
```





## Install

```PowerShell
PS C:\Users\genio> psperl -Install perl64-5.30.0.1
PS C:\Users\genio> psperl -Install perl32w64int-5.30.0.1
PS C:\Users\genio> psperl -Install perl32-5.30.0.1
```

Once you've selected which version of Perl you'd like to use from the list supplied in the `Available` command, you'd just need to install it.

This command will download the Portable Perl zip file from Strawberry Perl's site and store it locally in the `_zips` directory. It will then check that we got the expected file size and SHA1 checksum match. If it's not the right size or checksum, we'll remove the zip file and fail.

Given that we have the right file, we'll extract the zip file into your `_perls` directory. If we're installing `perl32w64int-5.30.0.1` then the path to that Perl would be `psperl_home_dir\_perls\perl32w64int-5.30.0.1`. So, if you installed PSPerl in `C:\Users\genio\_psperl`, the full path to that directory would be `C:\Users\genio\_psperl\_perls\perl32w64int-5.30.0.1`.

Just installing the version of Perl does _not_ make it ready for you to use immediately. This is simply because you may not yet want to work on that version. If you're using v5.30 and wanted to install another version, say v5.28, to test with, you can install that new version but you'd still be on v5.30 until you decide to change it with the `Use` or `Switch` commands.

## List

```PowerShell
PS C:\Users\genio> psperl -List # Shows you a list of the Perls you have
```

The `List` command will show you a listing of which Perls you currently have installed.


### Switch

```PowerShell
PS C:\Users\genio> psperl -Switch perl64-5.30.0.1
PS C:\Users\genio> psperl -Switch perl32w64int-5.30.0.1
PS C:\Users\genio> psperl -Switch perl32-5.30.0.1
```

The `Switch` command can be used to change your current environment setup to use any version of Perl we've already installed. This will be made persistent as we'll store the current version in our local config directory and every time you start a PowerShell session, we'll setup your environment for the selected version of Perl.

You can change which Perl you're using as many times as you'd like during a session, though you can only use one Perl at a time.

### Use

```PowerShell
PS C:\Users\genio> psperl -Use perl64-5.30.0.1
PS C:\Users\genio> psperl -Use perl32w64int-5.30.0.1
PS C:\Users\genio> psperl -Use perl32-5.30.0.1
```

The `Use` command works almost exactly as the `Switch` command does, but it isn't persistent. With this command, we just start using a different version of Perl, but we haven't switched our preference to that one. If my current preference is Perl v5.28 and it's what I have setup to start on my PowerShell sessions, I could temporarily use Perl v5.30 with this command. If I close my session and re-open it, I'll be back on Perl v5.28.

You can change which Perl you're using as many times as you'd like during a session, though you can only use one Perl at a time.

# AUTHOR

Chase Whitener `<capoeirab@cpan.org>`

# COPYRIGHT & LICENSE

Copyright 2019, Chase Whitener, All Rights Reserved.

You may use, modify, and distribute this package under the
same terms as Perl itself.


Changes
Changes log for PSPerl

0.1.2       2021

0.1.1       2021-04-14
  * Removed TravisCI testing
  * Altered the Setup() method to create the directory for the
    $Profile if it doesn't yet exist
  * Created a GitHub Workflow for testing purposes
  * Reverted to https://strawberryperl.com/releases.json

0.1.0       2020-06-29
  * Switch to https://strawberry.perl.bot/releases.json
    - Gets us HTTPS, and it's faster than strawberryperl.com
  * Add a -Version or -V command to show version information
  * Begin some structure for creating a -Local or -Lib command
    to deal with local::lib
  * Get some tags/releases back-filled
  * Added the Changes log

0.0.2       2020-04-30
  * Store the current version we're running in the environment
  * Fix for non-English Windows installs (thanks xenu)
  * Fix for error handling (thanks xenu)

0.0.1       2020-03-18
  * Initial Release:
  * psperl -Available
  * psperl -Init
  * psperl -List
  * psperl -Install
  * psperl -Use
  * psperl -Switch


src/PSPerl.ps1
class PSPerl {
    hidden [String] $URL;
    hidden [String] $rootPath;
    hidden [String] $profilePath;
    hidden [System.Object] $settings;

    # Constructor
    PSPerl([string]$rootPath, [string]$profilePath) {
        if(![System.IO.Directory]::Exists($rootPath)) {
            throw "Invalid root path provided";
        }
        $this.URL = 'https://strawberryperl.com/releases.json';
        $this.rootPath = $rootPath;
        $this.profilePath = $profilePath;
        $this.LoadSettings();
        $this.SaveSettings();
    }

    # Return the operating system's bitness. will be 64 or 32
    [int] ArchOS() {
        $the_bits = (Get-CimInstance Win32_OperatingSystem).OSArchitecture;
        if ($the_bits -match '64') { return 64; }
        elseif ($the_bits -match '32') { return 32; }
        throw "Invalid OS Architecture? We don't know what to do with a `"$($the_bits)`" architecture";
    }

    # return the current PowerShell's bitness. will be 64 or 32. might not be the same as the OS
    [int] ArchPS() {
        $the_bits = [System.Runtime.InterOpServices.Marshal]::SizeOf([System.IntPtr]::Zero)*8;
        if ($the_bits -eq '64') { return 64; }
        elseif ($the_bits -eq '32') { return 32; }
        throw "Invalid PowerShell Architecture? We don't know what to do with a $($the_bits)-bit PowerShell";
    }

    # get an array of available Portable Perl versions
    [Array] AvailablePerls() {
        [String]$cache_file = "$($this.rootPath)\_config\_release_cache.json";
        [Array]$all_portables = @();
        # check to see if we have a cached copy of the data first
        if (Test-Path $cache_file -NewerThan (Get-Date).AddHours(-1).AddMinutes(-30)) {
            $all_portables = (Get-Content -Raw -Path $cache_file | ConvertFrom-Json)
            return $all_portables;
        }

        # Ensures that Invoke-WebRequest uses TLS 1.2
        [Net.ServicePointManager]::SecurityProtocol = [Net.SecurityProtocolType]::Tls12
        $res = Invoke-RestMethod -Uri $this.URL -Method Get -Headers @{'Accept'='Application/json'}

        foreach ($row in $res) {
            # only use the info for the ones with the portable edition
            if ([bool]($row.PSobject.Properties.Name -contains 'edition') -And [bool]($row.edition.PSObject.Properties.Name -contains 'portable')) {
                # move edition info into the base object
                $row | Add-Member -Name "sha1" -Type NoteProperty -Value $row.edition.portable.sha1;
                $row | Add-Member -Name "sha256" -Type NoteProperty -Value $row.edition.portable.sha256;
                $row | Add-Member -Name "size" -Type NoteProperty -Value $row.edition.portable.size;
                $row | Add-Member -Name "url" -Type NoteProperty -Value $row.edition.portable.url;
                $row.PSObject.Properties.Remove('edition');
                # add some extra object members for version lookup simplicity later
                [int]$perl, [int]$major, [int]$minor, [int]$rel = $row.version.split('.');
                $row | Add-Member -Name "Perl" -Type NoteProperty -Value $perl;
                $row | Add-Member -Name "major" -Type NoteProperty -Value $major;
                $row | Add-Member -Name "minor" -Type NoteProperty -Value $minor;
                $row | Add-Member -Name "rel" -Type NoteProperty -Value $rel;
                # add an x64 and a 'USE_64_BIT_INT' field that's true/false
                $row | Add-Member -Name "x64" -Type NoteProperty -Value $false;
                $row | Add-Member -Name "USE_64_BIT_INT" -Type NoteProperty -Value $false;
                $row | Add-Member -Name "install_name" -Type NoteProperty -Value "perl32-$($row.version)";
                if ($row.archname -clike "*x64*") {
                    $row.x64 = $true;
                    $row.install_name = "perl64-$($row.version)";
                }
                else {
                    if ($row.archname -clike "*-64int") {
                        $row.USE_64_BIT_INT = $true;
                        $row.install_name = "perl32w64int-$($row.version)";
                    }
                }
                $all_portables += , $row;
            }
        }
        # on a 64-bit OS, we can use either 32- or 64-bit Portables
        if ($this.ArchOS() -eq 32) {
            # on a 32-bit OS, filter out all 64-bit Portables
            $all_portables = @(,($all_portables | Where-Object {-Not $_.x64}))
        }
        # save our data to the cache file
        $all_portables | ConvertTo-Json -depth 100 -Compress | Out-File $cache_file
        return $all_portables;
    }

    [void] ClearEnvironment() {
        # Write-Output "Cleaning your environment of any and all Perl variables";
        # remove any set Perl/local::lib variables
        if (Test-Path 'env:TERM') { Remove-Item env:\TERM }
        if (Test-Path 'env:PERL_JSON_BACKEND') { Remove-Item env:\PERL_JSON_BACKEND }
        if (Test-Path 'env:PERL_YAML_BACKEND') { Remove-Item env:\PERL_YAML_BACKEND }
        if (Test-Path 'env:PERL5LIB') { Remove-Item env:\PERL5LIB }
        if (Test-Path 'env:PERL5OPT') { Remove-Item env:\PERL5OPT }
        if (Test-Path 'env:PERL_MM_OPT') { Remove-Item env:\PERL_MM_OPT }
        if (Test-Path 'env:PERL_MB_OPT') { Remove-Item env:\PERL_MB_OPT }
        if (Test-Path 'env:PERL_LOCAL_LIB_ROOT') { Remove-Item env:\PERL_LOCAL_LIB_ROOT }
        if (Test-Path 'env:PSPERL_CURRENT') { Remove-Item env:\PSPERL_CURRENT }

        # Write-Output "Path to PSPerl is: $($this.rootPath)"
        # Go through the PATH and remove Perl-related items
        $good_array = @();
        $array = $env:Path.split(";", [System.StringSplitOptions]::RemoveEmptyEntries) | Select-Object -uniq
        Foreach ($item in $array) {
            if (-not $item.StartsWith($($this.rootPath))) {
                $good_array += ,$item;
            }
        }
        $env:Path = $good_array -join ';'
        return;
    }

    [void] Init() {
        $this.ClearEnvironment();
        $env:Path = "$($this.rootPath);$env:Path";
        if ($this.settings.current) { $this.Use($this.settings.current, $false); }
        return;
    }

    [void] Install([System.Object]$perl_obj) {
        Write-Host("Attempting to install $($perl_obj.install_name)");
        [String]$pathDir = "$($this.rootPath)\_perls\$($perl_obj.install_name)";
        if ([System.IO.Directory]::Exists($pathDir)) {
            Write-Host("");
            Write-Host("Perl $($perl_obj.install_name) is already installed. Try using it with:");
            Write-Host("");
            Write-Host("    psperl -use $($perl_obj.install_name)     # temporary. this session");
            Write-Host("    psperl -switch $($perl_obj.install_name)  # permanent. this and every subsequent session");
            Write-Host("");
            return;
        }

        [String]$pathZip = "$($this.rootPath)\_zips\$($perl_obj.install_name).zip";
        if (![System.IO.File]::Exists($pathZip)) {
            # 150743342
            Write-Host("Downloading $($perl_obj.url). This may take some time as it's $($perl_obj.size) bytes.");
            Invoke-WebRequest -Uri $perl_obj.url -OutFile $pathZip
        }

        # we SHOULD now have the file
        if ([System.IO.File]::Exists($pathZip)) {
            # check to make sure we got the right size file
            [int]$size = (Get-Item $pathZip).length;
            if ($size -ne $perl_obj.size) {
                Remove-Item -Path $pathZip -Force;
                throw "The file we have is $($size) bytes but we expected $($perl_obj.size). Deleting the file.";
            }
            # check the SHA1 checksums
            [String]$checksum = (Get-FileHash -Path $pathZip -Algorithm SHA1).hash;
            if ($checksum -ne $perl_obj.sha1) {
                Remove-Item -Path $pathZip -Force;
                throw "The file's SHA1 checksum is off. Deleting the file.";
            }
        }
        else {
            throw "We tried to download the file, but we didn't get it.";
        }

        # extract the zip into the directory
        Write-Host("Found $($pathZip) with the correct size and SHA1 checksum. Extracting.");
        $this.ExtractArchive($pathZip, $pathDir);
        Write-Host("");
        Write-Host("Installed $($perl_obj.install_name) in $($pathDir). Try using it with:");
        Write-Host("");
        Write-Host("    psperl -use $($perl_obj.install_name)     # temporary. this session");
        Write-Host("    psperl -switch $($perl_obj.install_name)  # permanent. this and every subsequent session");
        Write-Host("");
    }

    [void] ExtractArchive([String]$archive_path, [String]$out_dir) {
        # tar.exe from Cygwin or MSYS may struggle with Windows-style paths so
        # ensure we are using the one that came with Windows 10
        $tar_path = "$env:WINDIR\system32\tar.exe";
        if ([System.IO.File]::Exists($tar_path)) {
            [void](New-Item -ItemType Directory -Path $out_dir -Force);
            & $tar_path -xf $archive_path -C $out_dir;
            if (-Not $?) {
                Remove-Item -Recurse -Force $out_dir;
                throw "Unable to extract the archive.";
            }
        }
        else {
            # Expand-Archive is much slower than tar, so we only use it as a
            # fallback
            Expand-Archive $archive_path -DestinationPath $out_dir;
        }
    }

    [array] InstalledPerls() {
        return [Array]((Get-ChildItem -dir "$($this.rootPath)\_perls").Name);
    }

    # Load in some info about what it is we're doing here.
    [void] LoadSettings() {
        # start out here
        $this.settings = New-Object -TypeName PSCustomObject -Property @{
            current = ''
            is_on = $true
        };
        # we could be attempting to read settings prior to setup
        if (![System.IO.Directory]::Exists("$($this.rootPath)\_config")) {
            return;
        }

        [String]$settingsPath = "$($this.rootPath)\_config\settings.json";
        if (Test-Path $settingsPath) {
            [System.Object]$temp = (Get-Content -Raw -Path $settingsPath | ConvertFrom-Json);
            if ($temp.PSobject.Properties.Name -contains 'current') { $this.settings.current = [string]$temp.current; }
            if ($temp.PSobject.Properties.Name -contains 'is_on') { $this.settings.is_on = [bool]$temp.is_on; }
        }
        if (($this.settings.current) -And (-Not $this.InstalledPerls().Contains($this.settings.current))) {
            $this.settings.current = '';
        }
        return;
    }

    # Load in some info about what it is we're doing here.
    [void] SaveSettings() {
        # we could be attempting to save settings prior to setup
        if (![System.IO.Directory]::Exists("$($this.rootPath)\_config")) {
            return;
        }
        [String]$settingsPath = "$($this.rootPath)\_config\settings.json";
        $this.settings | ConvertTo-Json -depth 100 | Out-File $settingsPath
        return;
    }

    [void] Setup() {
        # check to ensure the $this.profilePath exists
        if (![System.IO.File]::Exists($this.profilePath)) {
            if (![System.IO.Directory]::Exists($this.profilePath)) {
                $profileDir = (Split-Path -parent $this.profilePath)
                Write-Host("You don't yet have the profile directory $($profileDir). We'll create it now.");
                New-Item -ItemType Directory $profileDir -ErrorAction SilentlyContinue
            }
            Write-Host("You don't yet have a profile at $($this.profilePath). We'll set that up now.");
            # this is like touch filename in linux
            New-Item -ItemType file $this.profilePath
        }
        $init_content = "# PSPerl Initialize`r`n& $($this.rootPath)\psperl.ps1 -Init"
        $found = Select-String -Quiet -Pattern "^# PSPerl Init" -Path $this.profilePath
        if (-not $found) {
            Add-Content -Path $this.profilePath -Value "`r`n$init_content"
        }

        # check to ensure the config directory exists
        if(![System.IO.Directory]::Exists("$($this.rootPath)\_config")) {
            Write-Host("You don't yet have a configuration directory.");
            # this is like touch filename in linux
            New-Item -ItemType directory "$($this.rootPath)\_config"
        }
        # check to ensure the \_zips directory exists
        if(![System.IO.Directory]::Exists("$($this.rootPath)\_zips")) {
            # this is like touch filename in linux
            New-Item -ItemType directory "$($this.rootPath)\_zips"
        }
        # check to ensure the config\perls directory exists
        if(![System.IO.Directory]::Exists("$($this.rootPath)\_perls")) {
            # this is like touch filename in linux
            New-Item -ItemType directory "$($this.rootPath)\_perls"
        }

        Write-Host("Your profile is located at:  $($this.profilePath)");
        Write-Host("Your config directory is at: $($this.rootPath)\_config");
        return;
    }

    [void] Use([string]$perl_install, [bool]$persistent = $false) {
        if (-Not $this.InstalledPerls().Contains($perl_install)) {
            throw "$($perl_install) isn't yet installed. Try installing it.";
        }
        $this.ClearEnvironment();
        $env:PSPERL_CURRENT = $perl_install;
        $env:PATH = "$($this.rootPath);$($env:PATH)";
        # which perl will we be using?
        [string]$path = "$($this.rootPath)\_perls\$($perl_install)";

        $env:PATH = "$($path)\perl\site\bin;$($path)\perl\bin;$($path)\c\bin;$($env:PATH)";

        # setup local::lib stuff
        # [String]$lib_path = "$($this.rootPath)\_libs\$($perl_install)\$($lib_name)";
        # $lib_path = $lib_path.Replace("\", "/");
        # $env:PATH = "$($lib_path)/bin;$($env:PATH)";
        # perl -Mlocal::lib="$($lib_path)" *>$null;
        # $env:PERL5LIB = "$($lib_path)/lib/perl5";
        # $env:PERL_LOCAL_LIB_ROOT= "$($lib_path)";
        # $env:PERL_MB_OPT = $("--install_base `"$($lib_path)`"");
        # $env:PERL_MM_OPT = $("INSTALL_BASE=$($lib_path)");
        if ($persistent) {
            $this.settings.current = $perl_install;
            $this.SaveSettings();
        }
    }
}


helpers/functions/Setup.ps1
<#
.SYNOPSIS
    Setup PSPerl

.DESCRIPTION
    Ensure the config directory is there. Ensure we have PSPerl setup in our $Profile.

#>
function Setup {
    Param
    (
        [Parameter(Mandatory=$true, Position=0)]
        [string] $psperl_path
    )
    # check to ensure the $Profile exists
    if(![System.IO.File]::Exists($Profile)) {
        Write-Output "You don't yet have a profile at $Profile. We'll set that up now.";
        # this is like touch filename in linux
        New-Item -ItemType file $Profile
    }
    $init_content = "# PSPerl Initialize`r`n& $psperl_path\psperl.ps1 -Init"
    $found =  Select-String -Quiet -Pattern "^# PSPerl Init" -Path $Profile
    if (-not $found) {
        Add-Content -Path $Profile -Value "`r`n$init_content"
    }

    # check to ensure the config directory exists
    if(![System.IO.Directory]::Exists("$psperl_path\_config")) {
        Write-Output "You don't yet have a configuration directory.";
        # this is like touch filename in linux
        New-Item -ItemType directory "$psperl_path\_config"
    }

    Write-Output "Your profile is located at:  $Profile"
    Write-Output "Your config directory is at: $psperl_path\_config"
    Init $psperl_path;
}


psperl.ps1
# command line arguments
param (
    [switch]$available = $false,
    [switch]$init = $false,
    [switch]$setup = $false,
    [switch]$list = $false,
    [switch]$version = $false,
    [switch]$v = $false,
    [string]$install = '',
    [string]$switch = '',
    [string]$use = '',
    [int]$major = 0
)

# Make sure we only attempt to work for PowerShell v5 and greater
# this allows the use of classes.
if ($PSVersionTable.PSVersion.Major -lt 5) {
    throw "PowerShell v5.0+ is required for psperl. https://docs.microsoft.com/en-us/powershell/scripting/setup/installing-windows-powershell?view=powershell-6";
}
$psperl_path = (Split-Path -parent $MyInvocation.MyCommand.Definition);
$env:PSPERL_ROOT = $psperl_path;
$env:PSPERL_VERSION = '0.1.2';

# Import our classes
. "$($psperl_path)\src\PSPerl.ps1";
$psperl = [PSPerl]::new($psperl_path, $Profile);

$global:DebugPreference = 'Continue'
if ($setup) { $psperl.Setup(); }
elseif ($init) { $psperl.Init(); }
elseif ($available) {
    [int]$ps_bits = $psperl.ArchPS();
    [int]$os_bits = $psperl.ArchOS();
    Write-Host("We're using a $($ps_bits)-bit PowerShell on a $($os_bits)-bit Windows OS.");
    Write-Host("On 64-bit Windows:");
    Write-Host("    32-bit PowerShell: You can use 32-bit, 32-bit USE_64_BIT_INT Perls without problem.");
    Write-Host("                       You _can_ use 64-bit Perls, but may run into trouble. RISK");
    Write-Host("    64-bit PowerShell: You can use 32-bit, 32-bit USE_64_BIT_INT, or 64-bit Perls.");
    Write-Host("On 32-bit Windows:");
    Write-Host("    You can only use 32-bit or 32-bit with USE_64_BIT_INT Perls");
    Write-Host("");
    [array]$data = $psperl.AvailablePerls();
    if ($major -gt 0) {
        Write-Host("Perls available where major version is: $($major)");
        Write-Host("");
        if ($os_bits -eq 64) {
            [String]$message = '64-bit Perls';
            if ($ps_bits -ne 64) { $message = "$($message) (RISKY on a 32-bit PS)"}
            Write-Host("$($message):");
            ForEach($row in ($data | Where-Object {($_.x64) -And ($_.major -eq $major)})) {
                Write-Host("   - $($row.install_name)");
            }
        }
        Write-Host("");
        Write-Host("32-bit with USE_64_BIT_INT Perls:")
        ForEach($row in ($data | Where-Object {(-Not $_.x64) -And ($_.USE_64_BIT_INT) -And ($_.major -eq $major)})) {
            Write-Host("   - $($row.install_name)");
        }
        Write-Host("");
        Write-Host("32-bit Perls:")
        ForEach($row in ($data | Where-Object {(-Not $_.x64) -And (-Not $_.USE_64_BIT_INT) -And ($_.major -eq $major)})) {
            Write-Host("   - $($row.install_name)");
        }
    }
    else {
        Write-Host("The current release for each major Perl:");
        Write-Host("");
        if ($os_bits -eq 64) {
            [String]$message = '64-bit Perls';
            if ($ps_bits -ne 64) { $message = "$($message) (RISKY on a 32-bit PS)"}
            Write-Host("$($message):");
            ForEach($row in ($data | Where-Object {($_.x64)} | Group-Object -Property major | Sort-Object -Descending -Property Name)) {
                Write-Host("   - $($row | ForEach-Object {$_.Group[0].install_name})");
            }
        }
        Write-Host("");
        Write-Host("32-bit with USE_64_BIT_INT Perls:")
        ForEach($row in ($data | Where-Object {(-Not $_.x64) -And ($_.USE_64_BIT_INT)} | Group-Object -Property major | Sort-Object -Descending -Property Name)) {
            Write-Host("   - $($row | ForEach-Object {$_.Group[0].install_name})");
        }
        Write-Host("");
        Write-Host("32-bit Perls:")
        ForEach($row in ($data | Where-Object {(-Not $_.x64) -And (-Not $_.USE_64_BIT_INT)} | Group-Object -Property major | Sort-Object -Descending -Property Name)) {
            Write-Host("   - $($row | ForEach-Object {$_.Group[0].install_name})");
        }
    }
    Write-Host("");
}
elseif ($install) {
    Write-Host("You'd like to install version $($install). Let's see what we can do.");
    [array]$data = $psperl.AvailablePerls();

    # grep through our available perls for the one specified by name.
    # this "grep" (Where-Object) is insane. If only one thing is found, it doesn't
    # return a list with one item. It just returns the item. If more than one are
    # found, then you get a list of items. -sigh
    # List form: $found.GetType() = 'System.Object[]'
    # Single Object form: $found.GetType() = 'System.Management.Automation.PSCustomObject'
    $found = $data | Where-Object {$_.install_name -eq $install};

    # if nothing is found, a null thingy is returned. This can be tested with a simple truthiness test
    if (-Not $found) {
        throw "No installable version of Perl found by the name $($install). Try `"psperl -available`" to get a list of available perl installations.";
    }
    # check to see if we got just one.
    if ($found -is 'System.Object') {
        $psperl.Install($found);
    }
    else {
        throw "Unexpected results `"$($found.GetType())`" searching for `"$($install)`". Try `"psperl -available`" to get a list of available perl installations.";
    }
}
elseif ($switch) { $psperl.Use($switch, $true); }
elseif ($use) { $psperl.Use($use, $false); }
elseif ($list) {
    Write-Host("Perls installed on your system: ");
    Write-Host("");
    ForEach ($dir in (Get-ChildItem -Path "$($env:PSPERL_ROOT)\_perls" -Directory)) {
        if ($dir.Name -eq $env:PSPERL_CURRENT) {
            Write-Host -NoNewline " * ";
        }
        else {
            Write-Host -NoNewline "   ";
        }

        Write-Host $dir.Name;
    }
}
elseif ($version -or $v) {
    Write-Host("This is PSPerl v$($env:PSPERL_VERSION)");
    Write-Host("");
    Write-Host("Directories of note:");
    Write-Host("$($env:PSPERL_ROOT)\_config");
    Write-Host("$($env:PSPERL_ROOT)\_locals");
    Write-Host("$($env:PSPERL_ROOT)\_perls");
    Write-Host("$($env:PSPERL_ROOT)\_zips");
    Write-Host("");
    Write-Host("https://github.com/genio/psperl");
    Write-Host("");
}
else {
    Get-Content -Path "$($env:PSPERL_ROOT)\helpfile.txt" | Write-Host
}
# To turn on Debugging, $global:DebugPreference = 'Continue'
# To turn off Debugging, $global:DebugPreference = 'SilentlyContinue'

exit;
